builtin/credential/ldap/backend.go

```diff
@@ -100,40 +100,44 @@ func (b *backend) Login(req *logical.Request, username string, password string)
 		return nil, logical.ErrorResponse("invalid connection returned from LDAP dial"), nil
 	}
 
-	bindDN, err := getBindDN(cfg, c, username)
+	bindDN, err := b.getBindDN(cfg, c, username)
 	if err != nil {
 		return nil, logical.ErrorResponse(err.Error()), nil
 	}
 
+	b.Logger().Printf("[DEBUG] auth/ldap: BindDN for %s is %s", username, bindDN)
+
+	// Try to bind as the login user. This is where the actual authentication takes place.
 	if err = c.Bind(bindDN, password); err != nil {
 		return nil, logical.ErrorResponse(fmt.Sprintf("LDAP bind failed: %v", err)), nil
 	}
 
-	userDN, err := getUserDN(cfg, c, bindDN)
+	userDN, err := b.getUserDN(cfg, c, bindDN)
 	if err != nil {
 		return nil, logical.ErrorResponse(err.Error()), nil
 	}
 
-	ldapGroups, err := getLdapGroups(cfg, c, userDN, username)
+	ldapGroups, err := b.getLdapGroups(cfg, c, userDN, username)
 	if err != nil {
 		return nil, logical.ErrorResponse(err.Error()), nil
 	}
+	b.Logger().Printf("[DEBUG] auth/ldap: Server returned %d groups", len(ldapGroups))
 
 	ldapResponse := &logical.Response{
 		Data: map[string]interface{}{},
 	}
 	if len(ldapGroups) == 0 {
 		errString := fmt.Sprintf(
-			"no LDAP groups found in userDN '%s' or groupDN '%s';only policies from locally-defined groups available",
-			cfg.UserDN,
+			"no LDAP groups found in groupDN '%s';only policies from locally-defined groups available",
 			cfg.GroupDN)
 		ldapResponse.AddWarning(errString)
 	}
 
 	var allGroups []string
 	// Import the custom added groups from ldap backend
 	user, err := b.User(req.Storage, username)
-	if err == nil && user != nil {
+	if err == nil && user != nil && user.Groups != nil {
+		b.Logger().Printf("[DEBUG] auth/ldap: adding %d local groups: %s\n", len(user.Groups), user.Groups)
```

Comment:
If you're doing to indicate what groups were added here, probably above the groups returned from the LDAP server should be indicated as well (rather than just the length). I'm fine either way, but consistency is good.